client/client.go

```diff
@@ -803,53 +780,148 @@ func (c *Client) Call(ctx context.Context, procedure string, options wamp.Dict,
+// To finish a progressive call the last callback MUST be called with isLast = true
+func (c *Client) CallProgressive(ctx context.Context, procedure string, options wamp.Dict, args wamp.List, kwargs wamp.Dict, finalrescb ResultHandler, progcb ProgressHandler) (SendProgressiveData, error) { //nolint:lll
+	if !c.Connected() {
+		return nil, ErrNotConn
+	}
+
+	if !c.sess.HasFeature(wamp.RoleDealer, wamp.FeatureProgressiveCalls) {
+		return nil, ErrProgCallNotSupportedByRouter
+	}
+
+	if options == nil {
+		options = wamp.Dict{}
+	}
+
+	options[wamp.OptProgress] = true
+
+	// If caller is willing to receive progressive results, create a channel to
+	// receive these on.  Then, start a goroutine to receive progressive
+	// results and call the callback for each.
+	var progChan chan *wamp.Result
+	var progDone chan struct{}
+	if progcb != nil {
+		options[wamp.OptReceiveProgress] = true
+		progChan = make(chan *wamp.Result)
+		progDone = make(chan struct{})
+		go func() {
+			for result := range progChan {
+				progcb(result)
 			}
+			close(progDone)
+		}()
+	}
+
+	if finalrescb == nil {
+		return nil, errors.New("no result handler provided")
+	}
+
+	id := c.idGen.Next()
+	c.expectReply(id)
+	message := &wamp.Call{
+		Request:   id,
+		Procedure: wamp.URI(procedure),
+		Options:   options,
+	}
+
+	err := c.prepareCallPayloadMessage(message, options, args, kwargs)
+
+	if err != nil {
+		return nil, err
+	}
+
+	c.sess.Send(message)
+
+	// As we can not wait for final result and must return control to the client code
+	// we wait for result in goroutine
+	go func() {
+		// Wait to receive RESULT message.
+		msg, err := c.waitForReplyWithCancel(ctx, id, procedure, progChan)
+
+		// Finish handling any remaining progressive results before returning the
+		// final result.
+		if progcb != nil {
+			close(progChan)
+			<-progDone
+		}
+
+		if err != nil {
+			finalrescb(nil, err)
+			return
+		}
+
+		switch msg := msg.(type) {
+		case *wamp.Result:
+
+			err := c.prepareCallResultMessage(msg)
 
 			if err != nil {
-				return nil, err
+				finalrescb(nil, err)
+				return
 			}
 
-			msg.Arguments = args
-			msg.ArgumentsKw = kwargs
+			finalrescb(msg, nil)
+			return
+		case *wamp.Error:
+			finalrescb(nil, RPCError{msg, procedure})
+			return
+		default:
+			finalrescb(nil, unexpectedMsgError(msg, wamp.RESULT))
+			return
+		}
+	}()
+
+	// now we need to return callback for feeding the progressive data
+	sendProgressiveData := func(options wamp.Dict, args wamp.List, kwargs wamp.Dict, isLast bool) error {
+		if options == nil {
+			options = wamp.Dict{}
 		}
-		return msg, nil
-	case *wamp.Error:
-		return nil, RPCError{msg, procedure}
-	default:
-		return nil, unexpectedMsgError(msg, wamp.RESULT)
+
+		options[wamp.OptProgress] = isLast
+
+		message := &wamp.Call{
+			Request:   id,
+			Procedure: wamp.URI(procedure),
+			Options:   options,
+		}
+
+		err := c.prepareCallPayloadMessage(message, options, args, kwargs)
```

Comment:
Nitpick, combine this with error checking from the lines below.

Reply:
O yeah! Forgot about that go idiomatic way :) Thnx!)